LVGL: multi-display support

JarmouniA opened the pull request:
LVGL multi-display support
Closes #83567

```
west build -p always -b native_sim/native/64 samples/modules/lvgl/multi_display
./build/zephyr/zephyr.exe
```
![Screenshot_lvgl2](https://github.com/user-attachments/assets/e29bcea4-af9a-4dae-bd7f-4a5f7c90dc71)

Tested for potential regressions with `samples/modules/lvgl/demos`, `samples/subsys/display/lvgl`, & `tests/lib/gui/lvgl` on native_sim_64 and STM32H750B_DK.

JarmouniA committed:
dts: bindings: add "zephyr,displays" compatible

Add "zephyr,displays" compatible for passing available display
controllers nodes to graphical libraries that have multi-display support
like LVGL

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

faxe1008 commented on modules/lvgl/include/lvgl_zephyr.h:
```diff
@@ -9,10 +9,25 @@
+
+#define LV_DISPLAYS_TO_INITIALIZE LISTIFY(DT_ZEPHYR_DISPLAYS_COUNT, LV_DISPLAY_IDX_MACRO, (,))
+
+#define LV_DISPLAY_EXTERN_OBJECTS(n) extern lv_display_t *lvgl_display_##n;
```
use `LVGL_DISPLAY_OBJ` in this macro?

faxe1008 commented on modules/lvgl/lvgl.c:
```diff
@@ -101,34 +113,31 @@ static void lvgl_log(lv_log_level_t level, const char *buf)
+		(lv_display_set_buffers(lv_display, &buf0_##n, &buf1_##n, BUFFER_SIZE(n),	   \
+		LV_DISPLAY_RENDER_MODE_PARTIAL);						   \
+	))                                              \
+	IF_DISABLED(CONFIG_LV_Z_DOUBLE_VDB,							   \
```
Maybe use COND_CODE0 so you can remove this IF_DISABLED?

faxe1008 commented on modules/lvgl/lvgl.c:
```diff
@@ -101,34 +113,31 @@ static void lvgl_log(lv_log_level_t level, const char *buf)
+	struct lvgl_disp_data *data =
+		(struct lvgl_disp_data *)lv_display_get_user_data(lv_display_p);
+	uint16_t hor_res = lv_display_get_horizontal_resolution(lv_display_p);
+	uint16_t ver_res = lv_display_get_vertical_resolution(lv_display_p);
```
Is there any benefit to renaming  the var?

faxe1008 commented on modules/lvgl/lvgl.c:
```diff
@@ -221,50 +230,53 @@ lv_result_t lv_mem_test_core(void)
-	lv_display_set_render_mode(display, LV_DISPLAY_RENDER_MODE_FULL);
-#endif
+	FOR_EACH(LV_DISPLAY_INIT, (), LV_DISPLAYS_TO_INITIALIZE)
+		;
```
Not needed space.

faxe1008 replied:
Hmm that is strange :D, I'd prefer living with the complaint and do a override on the merge.

fabiobaltieri replied:
just add a pair of `/* clang-format off */` `/* clang-format on */` marker, clang-format makes a mess with the macro quite often, no shame in formatting blocks manually

faxe1008 reviewed (commented):
Very nice, quick first impression :^)

JarmouniA committed:
modules: lvgl: add multi-display input support

Add support for binding LV input devs to multiple displays

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

JarmouniA commented:
> the touch input events for both screens are passed to the same screen, which is not ideal

@towen add in 777f8a1fb306581cb90f9847b85882e1f3fb4496

Could you test it by adding this to your overlay
```
	lvgl_pointer {
		compatible = "zephyr,lvgl-pointer-input";
		input = <&disp1_touch_controller>;
		display-idx = <0>;
		status = "okay";
	};

	lvgl_pointer2 {
		compatible = "zephyr,lvgl-pointer-input";
		input = <&disp2_touch_controller>;
		display-idx = <1>;
		status = "okay";
	};
```

faxe1008 commented on dts/bindings/input/zephyr,lvgl-common-input.yaml:
```diff
@@ -8,3 +8,10 @@ properties:
     type: phandle
     description: |
       Input device phandle.
+  display-idx:
```
I'd rather go for a phandle here since this will add verification

faxe1008 replied:
Also: I think this is also needed for all other input devices, right?

faxe1008 commented on samples/modules/lvgl/multi_display/boards/native_sim_native_64.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_SYS_HEAP_RUNTIME_STATS=y
+CONFIG_SERIAL=y
+CONFIG_UART_NATIVE_POSIX=y
+CONFIG_NO_OPTIMIZATIONS=y
```
Is there a reason for CONFIG_SERIAL, UART_NATIVE_POSIX and NO_OPTIMIZATIONS?

faxe1008 commented on samples/modules/lvgl/multi_display/prj.conf:
```diff
@@ -0,0 +1,22 @@
+CONFIG_LV_Z_MEM_POOL_SIZE=65536
+
+CONFIG_LV_USE_DEMO_WIDGETS=y
+CONFIG_LV_USE_DEMO_FLEX_LAYOUT=y
```
Not needed imo, since you are choosing the file in the cmakelists manually anyways. 
Thinking about this it might make sense to move the CMakeLists stuff from `samples/modules/lvgl/demo` to `modules/lvgl/CMakeLists.txt` so this can be pulled in for other samples (like this multidisplay one) more easily.
But as it stands this does nothing atm., so can be removed.

faxe1008 commented:
Amazing work @JarmouniA 🎉 , I guess the only thing missing to make this run smoothly inside of the simulator is the distinction in the SDL touch, so that each display is also linked to one instance, right?

faxe1008 commented:
@JarmouniA opened this PR which should fix the issue:
https://github.com/zephyrproject-rtos/zephyr/pull/87333/files

JarmouniA commented:
@faxe1008 rebasing this PR as is rn on top of #87333 breaks the 2nd window. Works again when `input-sdl-touch`/`lvgl-pointer-input` nodes, & `chosen/zephyr,touch` prop are removed from DTS (see https://github.com/JarmouniA/zephyr_dev/tree/test_sdl_multi-display-in). Something fishy is going on with SDL, may be related to #71410, will investigate further later.

Also, regarding https://github.com/zephyrproject-rtos/zephyr/pull/86815#discussion_r2002772130, when implemented (see https://github.com/JarmouniA/zephyr_dev/tree/dev_lvgl_multi_disp_idx-to-ph) it broke the 2nd display as well (without rebasing on top of #87333), super weird !!

faxe1008 commented:
I am pretty sure that this has to do something with the multithreaded usage of SDL.

faxe1008 commented:
@JarmouniA raised https://github.com/zephyrproject-rtos/zephyr/pull/87883 but this does not seem to resolve the issue with the second window not being rendered.

JarmouniA committed:
devicetree: macros: add display specific macros

Add macros to get display controllers phandles and count from
"zephyr,displays" compatible node

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

JarmouniA committed:
samples: modules: lvgl: add multi-display sample

Add LVGL multi-display sample

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

JarmouniA committed:
modules: lvgl: add multi-display support in Zephyr

Add multi-display support in Zephyr ontop of LVGL which already supports
it.

This change allows for creating buffers and structures for each display
statically and automatically from deviceTree info, and call LV init
routines for each display

Signed-off-by: Abderrahmane JARMOUNI <git@jarmouni.me>

pdgendt commented on samples/modules/lvgl/multi_display/README.rst:
```diff
@@ -0,0 +1,50 @@
+
+A sample showcasing LVGL multi-display support in Zephyr.
+
+By default, it runs Benchmark demo on the first display, and Widgets demo on the other ones (order as defined in the "displays" property of "zephyr,displays" compatible node in deviceTree).
+Which demos are run can be changed by modifying the value of CONFIG_LV_Z_DEMO_FIRST_DISP## and CONFIG_LV_Z_DEMO_OTHER_DISPS## Kconfig symbols.
```
Try to keep line lengths to a max of 100 characters in rst files.

JarmouniA commented on dts/bindings/display/zephyr,displays.yaml:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2025 Abderrahmane JARMOUNI
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Display controllers to pass to graphical libraries like LVGL.
+  Only one node with this compatible is allowed in the DeviceTree
+
+compatible: "zephyr,displays"
+
+properties:
+  displays:
+    type: phandles
+    required: true
+    description: |
+      Nodes of display controllers for graphical libraries to use
```
For some context, this was a suggestion from @mbolivar.
https://github.com/zephyrproject-rtos/zephyr/pull/84890#pullrequestreview-2585161312

decsny replied:
yeah, generally we are allowing a massive amount of technical debt in DT due to no proper multi instance static software configuration system in zephyr

faxe1008 commented on drivers/display/Kconfig.sdl:
```diff
@@ -94,4 +94,10 @@ config SDL_DISPLAY_THREAD_PRIORITY
 	help
 	  Drawing thread priority.
 
+config SDL_DISPLAY_MAX_WINDOW_LABEL
```
Hmm not against this but maybe should be a seperate PR?

JarmouniA replied:
Removed.

faxe1008 commented on drivers/display/display_sdl.c:
```diff
@@ -47,6 +47,7 @@ struct sdl_display_task {
 struct sdl_display_config {
 	uint16_t height;
 	uint16_t width;
+	const char label[CONFIG_SDL_DISPLAY_MAX_WINDOW_LABEL];
```
Also, why not have a const char* instead?

faxe1008 commented on drivers/display/display_sdl.c:
```diff
@@ -116,14 +117,13 @@ static void sdl_task_thread(void *p1, void *p2, void *p3)
+		&disp_data->texture, &disp_data->read_texture, &disp_data->background_texture,
+		CONFIG_SDL_DISPLAY_TRANSPARENCY_GRID_CELL_COLOR_1,
+		CONFIG_SDL_DISPLAY_TRANSPARENCY_GRID_CELL_COLOR_2,
+		CONFIG_SDL_DISPLAY_TRANSPARENCY_GRID_CELL_SIZE, config->label);
```
Regarding argument ordering, I'd put this further up in the list perhabs after  `&disp_data->window, dev,`

faxe1008 commented on drivers/display/display_sdl_bottom.h:
```diff
@@ -24,8 +24,8 @@ int sdl_display_init_bottom(uint16_t height, uint16_t width, uint16_t zoom_pct,
-			    uint32_t transparency_grid_color2,
-			    uint16_t transparency_grid_cell_size);
+			    uint32_t transparency_grid_color2, uint16_t transparency_grid_cell_size,
+			    const char *label);
```
formatting looks a bit odd.

faxe1008 commented on modules/lvgl/input/lvgl_common_input.c:
```diff
@@ -54,6 +55,11 @@ int lvgl_input_register_driver(lv_indev_type_t indev_type, const struct device *
+	if (common_data->disp == NULL) {
+		return -EINVAL;
+	}
+	lv_display_set_default(common_data->disp);
```
You want to use `lv_indev_set_display` instead here.

faxe1008 commented on modules/lvgl/input/lvgl_pointer_input.c:
```diff
@@ -129,6 +130,46 @@ static void lvgl_pointer_process_event(struct input_event *evt, void *user_data)
+	struct lvgl_disp_data *lv_disp_data;
+	lv_display_t *lv_disps[DT_ZEPHYR_DISPLAYS_COUNT];
+	lv_display_t *lv_disp = NULL;
+
```
Maybe this should be part of `lvgl_common_input.h` instead, because you also need to link the other pseudo devices to a certain screen. Something like:
```
lv_display_t* lvgl_input_get_display(const struct device* dev)
```

faxe1008 commented on modules/lvgl/lvgl.c:
```diff
@@ -21,58 +22,86 @@
+/* prevent unaligned memory accesses. */
+
+/* clang-format off */
+#define LV_DISPLAY_MACRO(n)                                                                        \
```
Nit naming, this relates to the display buffers, so should be part of it I guess.

faxe1008 commented on modules/lvgl/lvgl.c:
```diff
@@ -21,58 +22,86 @@
+
+FOR_EACH(LV_DISPLAY_MACRO, (), LV_DISPLAYS_IDX_LIST);
+
+#define BUFS_INIT_MACRO(n)                                                                         \
```
Naming: No need to abbreviate BUFFER and also adding MACRO as a suffix is a bit strange.

faxe1008 commented on samples/modules/lvgl/multi_display/Kconfig:
```diff
@@ -0,0 +1,89 @@
+	help
+	  Build flex layout demo application.
+
+config LV_Z_DEMO_SCROLL_FIRST_DISP
```
LVGL upstream has recently moved some of the newer demos to a new repository.
https://github.com/lvgl/lvgl/pull/8022

Since I do not see having a fork of that one as well or maintaining the sample in tree, as pseudo mirror, this sample will probably be removed in the next update of our LVGL fork. I mean it works but I guess it would make things slimer if we did not add it just to remove it a few weeks in the future.

faxe1008 commented on samples/modules/lvgl/multi_display/Kconfig:
```diff
@@ -0,0 +1,89 @@
+	bool "LVGL flex layout demo"
+	select LV_USE_DEMO_FLEX_LAYOUT
+	help
+	  Build flex layout demo application.
```
Same with this one, unfortunately.

JarmouniA commented:
@faxe1008 Thank you for your great feedback, I reworked the input part, I think it's much cleaner now.

faxe1008 commented on samples/modules/lvgl/multi_display/sample.yaml:
```diff
@@ -0,0 +1,22 @@
+    fixture: fixture_display
+    type: one_line
+    regex:
+      - "\\[\\w+ free bytes, \\w+ allocated bytes, overhead = \\w+ bytes | lvgl in malloc mode\\]"
```
Kinda divided on the whole sample with console harness. Had a larger discussion about this not too long ago. Personally I'd avoid adding it since things like the allocation printout is implementation detail inside of `sys_heap`. If someone was to change the text for whatever reason, he'd need to adjust the LVGL sample. That just feels odd.

faxe1008 commented:
Just also tested it. Very very nice, excited to get this in 💯

JarmouniA commented on modules/lvgl/input/lvgl_common_input.c:
```diff
@@ -42,13 +44,48 @@ static void lvgl_input_read_cb(lv_indev_t *indev, lv_indev_data_t *data)
+		}
+	}
+
+	LOG_ERR("LV display corresponding to display device %s not found", disp_dev->name);
```
Added `__ASSERT_NO_MSG` and kept the LOG_ERR so that, even if `CONFIG_ASSERT=n`, the user gets the error message.

uLipe reviewed (approved):
Tested also on my end, great work!

mmahadevan108 commented:
@JarmouniA Touch screen support stopped working on the FRDMRW612 board after this change was merged. Anything we need to do to update `drivers/input/input_gt911.c`  driver?
